cpp/ql/lib/semmle/code/cpp/models/interfaces/Throwing.qll

```diff
@@ -11,12 +11,71 @@ import semmle.code.cpp.models.Models
 import semmle.code.cpp.models.interfaces.FunctionInputsAndOutputs
 
 /**
- * A class that models the exceptional behavior of a function.
+ * Represents a type of exception,
+ * either Structured Exception Handling (SEH) or C++ exceptions.
  */
-abstract class ThrowingFunction extends Function {
+newtype TException =
+  /** Structured Exception Handling (SEH) exception */
+  TSehException() or
+  /** C++ exception */
+  TCxxException()
+
+/**
+ * Functions with information about how an exception is thrown or if one is thrown at all.
+ * If throwing details conflict for the same function, IR is assumed
+ * to use the most restricted interpretation, meaning taking options
+ * that stipulate no exception is raised, before the exception is always raised,
+ * before conditional exceptions.
+ *
+ * Annotations must specify if the exception is from SEH (structured exception handling)
+ * or ordinary c++ exceptions.
+ */
+abstract private class ExceptionAnnotation extends Function {
+  /**
+   * Returns the type of exception this annotation is for,
+   * either a CPP exception or a STructured Exception Handling (SEH) exception.
+   */
+  abstract TException getExceptionType();
+
+  /**
+   * Holds if the exception type of this annotation is for a Structured Exception Handling (SEH) exception.
+   */
+  final predicate isSeh() { this.getExceptionType() = TSehException() }
+
+  /**
+   * Holds if the exception type of this annotation is for a CPP exception.
+   */
+  final predicate isCxx() { this.getExceptionType() = TCxxException() }
+}
+
+/**
+ * A Function that is known to not throw an exception.
+ */
+abstract class NonThrowingFunction extends ExceptionAnnotation { }
+
+/**
+ * A function this is known to raise an exception.
+ */
+abstract class ThrowingFunction extends ExceptionAnnotation {
+  ThrowingFunction() { any() }
+
+  /**
+   * Holds if this function may raise an exception during evaluation.
+   * If `unconditional` is `false` the function may raise, and if `true` the function
+   * will always raise an exception.
+   * Do not specify `none()` if no exception is raised, instead use the
+   * `NonThrowingFunction` class instead.
+   */
+  abstract predicate raisesException(boolean unconditional);
+
+  /**
+   * Holds if this function will always raise an exception if called
+   */
+  final predicate alwaysRaisesException() { this.raisesException(true) }
+
   /**
-   * Holds if this function may throw an exception during evaluation.
-   * If `unconditional` is `true` the function always throws an exception.
+   * Holds if this function may raise an exception if called but
+   * it is not guaranteed to do so. I.e., the function does not always raise an exception.
    */
-  abstract predicate mayThrowException(boolean unconditional);
+  final predicate mayRaiseException() { this.raisesException(false) }
```

Comment:
Similarly, the `mayThrowException` will need to be kept for now and marked as deprecated.

Reply:
Fixed

Reply:
Need to rethink this generally for backwards compatibility despite my 'fix', stand by...

Reply:
Ok I decided to move back to the old mechanic. I switched because SEH doesn't really throw an exception per se. But it's fine, and easier to just keep it the way it was. The changes have been made, just waiting for the checks to pass. I saw some regressions that didn't make any sense, and I couldn't replicate them locally in a prior push, but as of this message a new run is in progress, so hopefully it is all set.

Reply:
I'm seeing failures with some security/CWE tests, but I cannot recreate that locally. @MathiasVP  is that a common discrepancy? can you maybe run the test locally as well? Thye all passed for me.